Content.Shared/Execution/ExecutionSystem.cs

```diff
@@ -0,0 +1,397 @@
+        var defaultExecutionInternal = DefaultInternalMeleeExecutionMessage;
+        var defaultExecutionExternal = DefaultExternalMeleeExecutionMessage;
+
+        if (HasComp<GunComponent>(weapon))
+        {
+            if(!CanExecuteWithGun(weapon, victim, attacker))
+                return;
+
+            defaultSuicideExternal = DefaultExternalGunSuicideMessage;
+            defaultSuicideInternal = DefaultInternalGunSuicideMessage;
+            defaultExecutionExternal = DefaultInternalGunExecutionMessage;
+            defaultExecutionInternal = DefaultExternalGunExecutionMessage;
+        }
+
+
+        string internalMsg;
+        string externalMsg;
+
+        if (attacker == victim)
+        {
+            internalMsg = comp.SuicidePopupInternal ?? defaultSuicideInternal;
+            externalMsg = comp.SuicidePopupExternal ?? defaultSuicideExternal;
+        }
+        else
+        {
+            internalMsg = comp.ExecutionPopupInternal ?? defaultExecutionInternal;
+            externalMsg = comp.ExecutionPopupExternal ?? defaultExecutionExternal;
+        }
+        ShowExecutionInternalPopup(internalMsg, attacker, victim, weapon);
+        ShowExecutionExternalPopup(externalMsg, attacker, victim, weapon);
+
+        var doAfter =
+            new DoAfterArgs(EntityManager, attacker, comp.DoAfterDuration, new ExecutionDoAfterEvent(), weapon, target: victim, used: weapon)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                BreakOnDamage = true,
+                NeedHand = true
+            };
+
+        _doAfterSystem.TryStartDoAfter(doAfter);
+
+    }
+
+    private bool CanExecuteWithAny(EntityUid victim, EntityUid attacker)
+    {
+        // No point executing someone if they can't take damage
+        if (!TryComp<DamageableComponent>(victim, out _))
+            return false;
+
+        // You can't execute something that cannot die
+        if (!TryComp<MobStateComponent>(victim, out var mobState))
+            return false;
+
+        // You're not allowed to execute dead people (no fun allowed)
+        if (_mobStateSystem.IsDead(victim, mobState))
+            return false;
+
+        // You must be able to attack people to execute
+        if (!_actionBlockerSystem.CanAttack(attacker, victim))
+            return false;
+
+        // The victim must be incapacitated to be executed
+        if (victim != attacker && _actionBlockerSystem.CanInteract(victim, null))
+            return false;
+
+        // All checks passed
+        return true;
+    }
+
+    private bool CanExecuteWithMelee(EntityUid weapon, EntityUid victim, EntityUid user)
+    {
+        if (!CanExecuteWithAny(victim, user))
+            return false;
+
+        // We must be able to actually hurt people with the weapon
+        if (!TryComp<MeleeWeaponComponent>(weapon, out var melee) && melee!.Damage.GetTotal() > 0.0f)
+            return false;
```

Comment:
This is going to cause a null exception, null suppression should be used sparingly.

Reply:
I thought I had removed all the null suppression code along with that specific if statement which should have its conditions inverted. I guess I accidentally reverted it somewhere? I'll have it removed tomorrow.